### PR TITLE
WIP ModelExportInfra

### PR DESCRIPTION
**Pull request by skanjila**
# Code Pull Requests

Please provide the following:

- a set of classes to do model export in onnx and coreml

# Documentation Pull Requests

Note that the documentation HTML files are in `docs/` while the Markdown sources are in `mkdocs/docs`.

If you are proposing a modification to the documentation you should change only the Markdown files.

`api.md` is automatically generated from the docstrings in the code, so if you want to change something in that file, first modify `ludwig/api.py` docstring, then run `mkdocs/code_docs_autogen.py`, which will create `mkdocs/docs/api.md` .

### CODE REVIEW

File: ludwig/model_export/coreml_exporter.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+from model import MyNet
+
+from ludwig.api import LudwigModel
+from ludwig.model_export.base_model_exporter import BaseModelExporter
+
+ class CoreMLExporter(BaseModelExporter):
+    def export_classifier(self, model_path, export_path):
+        ludwig_model = LudwigModel.load(model_path)
+        model = CoreMLExporter(ludwig_model.model)  # Wrap the model
+        #model.eval()  # inference mode, is this needed.. I think onnx export does this for us
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:14]: I don't think this line is needed

File: ludwig/model_export/coreml_exporter.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+        width = ludwig_model.config["input_features"][0]["preprocessing"]["width"]
+        height = ludwig_model.config["input_features"][0]["preprocessing"]["height"]
+        example_input = torch.randn(1, 3, width, height, requires_grad=True)
+
+        # Create the model and load the weights
+        model = MyNet()
+        model.load_state_dict(torch.load(model_path))
+        # Create dummy input
+        dummy_input = torch.rand(1, 3, 32, 32)
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:24]: How come we don't use `example_input` here? Am I missing somthing?

File: ludwig/model_export/coreml_exporter.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import torch
+from model import MyNet
+
+from ludwig.api import LudwigModel
+from ludwig.model_export.base_model_exporter import BaseModelExporter
+
+ class CoreMLExporter(BaseModelExporter):
+    def export_classifier(self, model_path, export_path):
+        ludwig_model = LudwigModel.load(model_path)
+        model = CoreMLExporter(ludwig_model.model)  # Wrap the model
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:13]: Is this `model` referenced anywhere? It is overwritten on line 21

File: ludwig/model_export/coreml_exporter.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        # Create the model and load the weights
+        model = MyNet()
+        model.load_state_dict(torch.load(model_path))
+        # Create dummy input
+        dummy_input = torch.rand(1, 3, 32, 32)
+        # Trace the model
+        traced_model = torch.jit.trace(model, dummy_input)
+        # Create the input image type
+        input_image = ct.ImageType(name=example_input, shape=(1, 3, 32, 32), scale=1/255)
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:28]: Nit: can we do `scale=1 / 255.0` to ensure that it is a float.

Also, I thought example input has `shape=(1, 3, width, height)`. Where did `32` come from?

File: ludwig/model_export/onnx_exporter.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        self.model = model
+
+    @abstractmethod
+    def forward(self, x):
+        return self.model({"image_path": x})
+
+    def export_classifier(self, model_path, export_path):
+        ludwig_model = LudwigModel.load(model_path)
+        model = OnnxExporter(ludwig_model.model)  # Wrap the model
+        model.eval()  # inference mode, is this needed.. I think onnx export does this for us
```
Comment (saad-palapa) [inline ludwig/model_export/onnx_exporter.py:19]: I think the `ONNX` exporter does the `eval()`. Can you double check?

File: ludwig/model_export/onnx_exporter.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+        width = ludwig_model.config["input_features"][0]["preprocessing"]["width"]
+        height = ludwig_model.config["input_features"][0]["preprocessing"]["height"]
+        example_input = torch.randn(1, 3, width, height, requires_grad=True)
+
+        torch.onnx.export(
+            model,
+            example_input,
+            export_path,
+            opset_version=18,
```
Comment (saad-palapa) [inline ludwig/model_export/onnx_exporter.py:29]: Should `opset_version` and other `onnx` args be an arg in `export_classifier`?

Maybe it can look like:
```
def export_classifier(self, model_path, export_path, onnx_export_args_override):
```

**Review by saad-palapa — CHANGES_REQUESTED** [review]

File: ludwig/model_export/coreml_exporter.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        ludwig_model = LudwigModel.load(model_path)
+        model = CoreMLExporter(ludwig_model.model)  # Wrap the model
+        #model.eval()  # inference mode, is this needed.. I think onnx export does this for us
+
+        width = ludwig_model.config["input_features"][0]["preprocessing"]["width"]
+        height = ludwig_model.config["input_features"][0]["preprocessing"]["height"]
+        example_input = torch.randn(1, 3, width, height, requires_grad=True)
+
+        # Create the model and load the weights
+        model = MyNet()
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:21]: Where does this come from?

File: ludwig/model_export/base_model_exporter.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    @abstractmethod
+    def forward(self, x):
+        pass
+
+    @abstractmethod
+    def export_classifier_onnx(model_path, export_path):
+        pass
+
+    @abstractmethod
+    def quantize(path_fp32, path_int8):
```
Comment (saad-palapa) [inline ludwig/model_export/base_model_exporter.py:21]: I think it's better design to call this:
```
def quantize(model_path, quantized_path):
```

The reason is that not all quantization is int8

File: ludwig/model_export/coreml_exporter.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+ class CoreMLExporter(BaseModelExporter):
+    def export_classifier(self, model_path, export_path):
+        ludwig_model = LudwigModel.load(model_path)
+        model = CoreMLExporter(ludwig_model.model)  # Wrap the model
+        #model.eval()  # inference mode, is this needed.. I think onnx export does this for us
+
+        width = ludwig_model.config["input_features"][0]["preprocessing"]["width"]
+        height = ludwig_model.config["input_features"][0]["preprocessing"]["height"]
+        example_input = torch.randn(1, 3, width, height, requires_grad=True)
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:18]: Should we have logic to check if this is` channels_first` or `channels_last`?

File: ludwig/model_export/coreml_exporter.py (hunk 2 above)
Comment (skanjila) [reply ludwig/model_export/coreml_exporter.py]: fixed

File: ludwig/model_export/coreml_exporter.py (hunk 1 above)
Comment (skanjila) [reply ludwig/model_export/coreml_exporter.py]: removed

File: ludwig/model_export/coreml_exporter.py (hunk 3 above)
Comment (skanjila) [reply ludwig/model_export/coreml_exporter.py:16]: its used in this line of code:  traced_model = torch.jit.trace(model, example_input)

File: ludwig/model_export/coreml_exporter.py (hunk 4 above)
Comment (skanjila) [reply ludwig/model_export/coreml_exporter.py]: fixed

File: ludwig/model_export/onnx_exporter.py (hunk 5 above)
Comment (skanjila) [reply ludwig/model_export/onnx_exporter.py]: removed

File: ludwig/model_export/onnx_exporter.py (hunk 6 above)
Comment (skanjila) [reply ludwig/model_export/onnx_exporter.py:28]: added

File: ludwig/model_export/base_model_exporter.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    @abstractmethod
+    def forward(self, x):
+        pass
+
+    @abstractmethod
+    def export_classifier(self, model_path, export_path, export_args_override):
+        pass
+
+    @abstractmethod
+    def quantize(self, path_fp32, path_int8):
```
Comment (saad-palapa) [inline ludwig/model_export/base_model_exporter.py:21]: The variable naming assumes all non-quantized models to be fp32 and quantized to be int8. This is not always the case. For example, we can have fp16 => fp4 (https://youtu.be/MK4k64vY3xo?si=ULlFZQa_2DSvcgpX&t=1970)

Perhaps the function signature should be:
```
def quantize(self, model_path, quantized_model_path)
```

File: ludwig/model_export/coreml_exporter.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        coreml_model_updated = ct.models.MLModel(spec)
+
+        # Save the CoreML model
+        coreml_model_updated.save(export_path)
+
+    @abstractmethod
+    def quantize(self, path_fp32, path_int8):
+        import coremltools.optimize.coreml as cto
+
+        ludwig_model = LudwigModel.load(path_fp32)
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:45]: For now this looks good. But can we add a `@todo` to add more quantization options.

File: ludwig/model_export/onnx_exporter.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+class OnnxExporter(ABC):
+    def __init__(self, model):
+        self.model = model
+
+    @abstractmethod
+    def forward(self, x):
+        return self.model({"image_path": x})
+
+    def export_classifier(self, model_path, export_path):
```
Comment (saad-palapa) [inline ludwig/model_export/onnx_exporter.py:16]: This is missing `export_args_override`

File: ludwig/model_export/base_model_exporter.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+class BaseModelExporter(ABC):
+    def __init__(self, model):
+        self.model = model
+
+    @abstractmethod
+    def forward(self, x):
+        pass
+
+    @abstractmethod
+    def export_classifier(self, model_path, export_path, export_args_override):
```
Comment (saad-palapa) [inline ludwig/model_export/base_model_exporter.py:17]: Can we remove this and only have `export_args_override` in `OnnxExporter`.

**Review by saad-palapa — APPROVED** [review]
Looks good. Left a few comments to consider.

File: ludwig/model_export/coreml_exporter.py (hunk 7 above)
Comment (skanjila) [reply ludwig/model_export/coreml_exporter.py]: Removed

File: ludwig/model_export/coreml_exporter.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    def export_classifier(self, model_path, export_path):
+        ludwig_model = LudwigModel.load(model_path)
+        model = CoreMLExporter(ludwig_model.model)  # Wrap the model
+
+        width = ludwig_model.config["input_features"][0]["preprocessing"]["width"]
+        height = ludwig_model.config["input_features"][0]["preprocessing"]["height"]
+        example_input = torch.randn(1, 3, width, height, requires_grad=True)
+
+        # Create the model and load the weights
+        ludwig_model.load_state_dict(torch.load(model_path))
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:23]: I tried running this and am getting this error:
```
Traceback (most recent call last):
  File "/palapa/auto-ml/export_classifier_coreml.py", line 82, in <module>
    export_classifier_coreml(model_id_from_args)
  File "/palapa/auto-ml/export_classifier_coreml.py", line 77, in export_classifier_coreml
    model.export_classifier(path, coreml_path)
  File "/palapa/auto-ml/export_classifier_coreml.py", line 38, in export_classifier
    ludwig_model.load_state_dict(torch.load(model_path))
AttributeError: 'LudwigModel' object has no attribute 'load_state_dict'
```

Here is the code I'm running it with:
```
def export_classifier_coreml(model_id):
    path = f"/palapa/auto-ml/saved_models/{model_id}/run/model"
    coreml_path = f"/palapa/auto-ml/saved_models/{model_id}/weights/best.mlmodel"

    ludwig_model = LudwigModel.load(path)
    model = CoreMLExporter(ludwig_model.model)

    model.export_classifier(path, coreml_path)
```

File: ludwig/model_export/onnx_exporter.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+import onnx
+import torch
+
+from ludwig.api import LudwigModel
+from ludwig.model_export.base_model_exporter import LudwigTorchWrapper
+
+
+class OnnxExporter(ABC):
+    def export_classifier(self, model_id, model_path, export_path, input_model_name, output_model_name):
```
Comment (saad-palapa) [inline ludwig/model_export/onnx_exporter.py:12]: Some of these params are not being used. It think the function signature should look like:

```
def export(self, model_path, export_path)
```

I don't think it's necessary to include `_classifier` in the function name

File: ludwig/model_export/coreml_exporter.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
+from abc import ABC, abstractmethod
+
+import coremltools as ct
+import torch
+
+from ludwig.api import LudwigModel
+from ludwig.model_export.base_model_exporter import BaseModelExporter
+
+
+class CoreMLExporter(BaseModelExporter):
```
Comment (saad-palapa) [inline ludwig/model_export/coreml_exporter.py:10]: Can we leave the CoreML stuff to another PR. Let's keep the task here small

File: ludwig/model_export/base_model_exporter.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+        return self.model({"image_path": x})
+
+
+class BaseModelExporter(ABC):
+    @abstractmethod
+    def export_classifier(self, model_path, export_path, export_args_override):
+        pass
+
+    @abstractmethod
+    def quantize(self, path_fp32, path_int8):
```
Comment (saad-palapa) [inline ludwig/model_export/base_model_exporter.py:23]: Can we defer teh `quantize` function to another PR. Let's keep this one small

File: ludwig/model_export/onnx_exporter.py [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+            example_input,
+            os.path.join(export_path, output_model_name),
+            opset_version=18,
+            export_params=True,
+            do_constant_folding=True,
+            input_names=["input"],
+            output_names=["combiner_hidden_1", "output", "combiner_hidden_2"],
+        )
+
+    def quantize_onnx(self, model_id, path_fp32, path_int8):
```
Comment (saad-palapa) [inline ludwig/model_export/onnx_exporter.py:32]: Can we leave quantization off to another PR?

File: ludwig/model_export/onnx_exporter.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+        from onnxruntime.quantization import quantize_dynamic
+
+        quantize_dynamic(path_fp32, path_int8)  # type: ignore
+
+    def check_model_export(self, model_id, export_path, output_model_name):
```
Comment (saad-palapa) [inline ludwig/model_export/onnx_exporter.py:37]: Can we change this function signature to:
```
def check_model_export(self, path)
```

I don't think it's necessary to separate folder and filename

**Review by saad-palapa — CHANGES_REQUESTED** [review]
